rfc/rfc-0000-input-output-correlation.md

```diff
@@ -0,0 +1,217 @@
+    output: "$(status.outputs.url)"             #evaluated against resource status
+  - input: "$(sources.source.revision)"         #evaluated against inputs in template context
+    output: "$(status.outputs.scannedRevision)" #evaluated against resource status
+```
+
+# How it Works
+
+[how-it-works]: #how-it-works
+
+When Cartographer processes a resource, the jsonpath expressions in each `observedMatches` `input` and `output` fields
+are evaluated and compared. If they are all equal, then the output is considered to be correlated with the input(s).
+
+When output cannot be correlated to input, Cartographer will not propagate the output forward. Doing so ensures we can
+always correlate outputs with inputs on subsequent resources, and also mitigates behavior of bad actors in the cluster
+as described
+in [Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556).
+
+For cases where the output does not contain necessary information to match against inputs, one may instead match against
+the spec of the resource:
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterImageTemplate
+metadata:
+  name: example-image-building-thing
+spec:
+  template:
+    apiVersion: my-builder/v1
+    kind: ImageBuilderThatDoesntExposeInputDetailsInItsOutput
+    metadata:
+      name: ...
+    spec:
+      sourceUrl: "$(sources.source.url)$"
+      sourceRevision: "$(sources.source.url)$"
+observedMatches:
+  - input: "$(sources.source.url)"              #evaluated against inputs in template context
+    output: "$(spec.sourceUrl)"                 #evaluated against resource spec
+  - input: "$(sources.source.revision)"         #evaluated against inputs in template context
+    output: "$(spec.sourceRevision)"            #evaluated against resource spec
+```
+
+Since the spec of the resource may change independently of the output, when the `outputMatches` makes any assertion
+against the `spec`, Cartographer must also ensure that the current output is a result of processing the current spec by
+making additional assertions:
+
+* Resource is healthy, as per [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+* Resource `observedGeneration == generation`
+
+If these assertions cannot be made, for example, if the resource does not implement the `observedGeneration` pattern,
+then Cartographer stops processing the resource and flags an error on it. Great care must also be taken that `healthy`
+status rules specified by the template author corresponds to the situation where the output is indeed a result of the
+spec with the current `observedGeneration`, otherwise outputs may not truly be the result of an input.
+
+Additionally, relying on the spec to make assertions imposes a further restriction to avoid stalling artifact
+propagation: Cartographer must not stamp out any further changes to the resource until it has finished processing the
```

Comment:
Above, the RFC states that
> Cartographer will not propagate artifacts output
which I refer to as _reading_ an object.
Here the RFC states
> Cartographer must not stamp out any further changes to the resource
which I refer to as _writing_ an object.

Is this proposal to hold reads or writes or both?

Reply:
Both.